velox/functions/lib/Re2Functions.cpp

```diff
@@ -809,71 +893,81 @@ std::vector<std::shared_ptr<exec::FunctionSignature>> re2ExtractSignatures() {
   };
 }
 
-std::pair<PatternKind, vector_size_t> determinePatternKind(StringView pattern) {
-  vector_size_t patternLength = pattern.size();
-  vector_size_t i = 0;
-  // Index of the first % or _ character.
-  vector_size_t wildcardStart = -1;
-  // Index of the first character that is not % and not _.
-  vector_size_t fixedPatternStart = -1;
-  // Total number of % characters.
-  vector_size_t anyCharacterWildcardCount = 0;
-  // Total number of _ characters.
-  vector_size_t singleCharacterWildcardCount = 0;
-  auto patternStr = pattern.data();
+PatternMetadata determinePatternKind(const StringView& pattern) {
+  auto i = 0;
+  const auto patternLength = pattern.size();
+  const auto patternString = pattern.str();
+  auto wildcardStart = 0;
+  auto numAnyWildcardStream = 0;
+  auto numWildcards = 0;
+  auto numAnyCharacterWildcards = 0;
+  auto numSingleCharacterWildcards = 0;
+  auto fixedPatternStart = 0;
+  auto numFixedPatterns = 0;
+  std::vector<std::string> fixedPatterns;
 
   while (i < patternLength) {
-    if (patternStr[i] == '%' || patternStr[i] == '_') {
-      // Ensures that pattern has a single contiguous stream of wildcard
-      // characters.
-      if (wildcardStart != -1) {
-        return std::make_pair(PatternKind::kGeneric, 0);
-      }
+    if (patternString[i] == '%' || patternString[i] == '_') {
       // Look till the last contiguous wildcard character, starting from this
       // index, is found, or the end of pattern is reached.
       wildcardStart = i;
       while (i < patternLength &&
-             (patternStr[i] == '%' || patternStr[i] == '_')) {
-        singleCharacterWildcardCount += (patternStr[i] == '_');
-        anyCharacterWildcardCount += (patternStr[i] == '%');
+             (patternString[i] == '%' || patternString[i] == '_')) {
+        numSingleCharacterWildcards += (patternString[i] == '_');
+        numAnyCharacterWildcards += (patternString[i] == '%');
         i++;
       }
+      numAnyWildcardStream += (numAnyCharacterWildcards > 0);
     } else {
-      // Ensure that pattern has a single fixed pattern.
-      if (fixedPatternStart != -1) {
-        return std::make_pair(PatternKind::kGeneric, 0);
-      }
       // Look till the end of fixed pattern, starting from this index, is found,
       // or the end of pattern is reached.
       fixedPatternStart = i;
       while (i < patternLength &&
-             (patternStr[i] != '%' && patternStr[i] != '_')) {
+             (patternString[i] != '%' && patternString[i] != '_')) {
         i++;
       }
+      numFixedPatterns++;
+      fixedPatterns.emplace_back(
+          patternString.substr(fixedPatternStart, i - fixedPatternStart));
     }
   }
+  numWildcards = numSingleCharacterWildcards + numAnyCharacterWildcards;
```

Comment:
You can drop the usage of Character in these names so numSingle, numAny is sufficient